File: src/main/java/com/miron4dev/dsa/algorithm/string/Anagram.java
package com.miron4dev.dsa.algorithm.string;

/**
 * Given two strings s and t, write a function to determine if t is an anagram of s.
 */
public class Anagram {

    private final String s;

    public Anagram(String s) {
        this.s = s;
    }

    public boolean isAnagram(String t) {
        if (s.length() != t.length()) {
            return false;
        }
        int[] temp = new int[26];
        for (char c : s.toCharArray()) {
            temp[c - 'a'] = temp[c - 'a'] + 1;
        }
        for (char c : t.toCharArray()) {
            temp[c - 'a'] = temp[c - 'a'] - 1;
        }
        for (int i : temp) {
            if (i != 0) {
                return false;
            }
        }
        return true;
    }
}

File: ruby/lib/algorithms/leetcode/string/anagram.rb
module Algorithms::Leetcode::String

  #
  # Given two strings str1 and str2, write a function to determine if str1 is an anagram of str2.
  #
  class Anagram

    OFFSET = 'a'.ord

    def anagram?(str1, str2)
      return false if str1.length != str2.length

      freq = Array.new(26, 0)

      str1.each_byte {|c| freq[c - OFFSET] = freq[c - OFFSET] + 1}
      str2.each_byte {|c| freq[c - OFFSET] = freq[c - OFFSET] - 1}

      freq.all?(&:zero?)
    end
  end
end
